File: java/src/main/java/linked_lists/SinglyLinkedList.java
package linked_lists;

import java.util.ArrayList;
import java.util.List;

public class SinglyLinkedList<T> {
    private SNode<T> head = null;
    private SNode<T> tail = null;

    static class SNode<T> {
        T data;
        SNode<T> next;

        public SNode(T data) {
            this.data = data;
        }
    }

    public void insertLast(SNode<T> sNode) {
        if (head == null && tail == null) {
            tail = head = sNode;
        } else {
            tail.next = sNode;
            tail = sNode;
        }
    }

    public void insertFirst(SNode<T> sNode) {
        if (head == null && tail == null) {
            tail = head = sNode;
        } else {
            sNode.next = head;
            head = sNode;
        }
    }

    public T[] getAsArray() {
        List<T> collection = new ArrayList<>();

        SNode<T> traversal = head;
        while (traversal != null) {
            collection.add(traversal.data);
            traversal = traversal.next;
        }

        return (T[]) collection.toArray();
    }

    public void insertAt(int index, SNode<T> sNode) {
        SNode<T> traversalFront = head;
        SNode<T> traversalBack = null;

        int pointer = 0;
        while (traversalFront != null) {

            if (pointer == index) {
                sNode.next = traversalFront;

                if (traversalBack != null) {
                    traversalBack.next = sNode;
                }

                break;
            }

            pointer++;
            traversalBack = traversalFront;
            traversalFront = traversalFront.next;
        }

        if (sNode.next == head) {
            head = sNode;
        }
    }

    public void clear() {
        SNode<T> traversal = head;
        while (traversal != null) {
            SNode<T> temp = traversal.next;
            traversal.next = null;
            traversal = temp;
        }

        tail = head = null;
    }

    public void removeFirst() {
        if (head == tail) {
            head = tail = null;
        } else {
            head = head.next;
        }
    }

    public void removeLast() {
        if (head == tail) {
            head = tail = null;
        } else {
            SNode<T> traversal = head;
            while (traversal != null) {
                if (traversal.next == tail) {
                    traversal.next = null;
                    tail = traversal;
                    break;
                }

                traversal = traversal.next;
            }
        }
    }
}

File: csharp/src/main/linked_lists/SinglyLinkedList.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace linked_lists {
    public class SinglyLinkedList<T> {
        public SNode Head { get; set; }

        public SNode Tail { get; set; }

        internal void insertFirst (SNode sNode) {
            if (Head == null && Tail == null) {
                Head = Tail = sNode;
            } else {
                sNode.Next = Head;
                Head = sNode;
            }
        }

        internal void insertLast (SNode sNode) {
            if (Head == null && Tail == null) {
                Head = Tail = sNode;
            } else {
                Tail.Next = sNode;
                Tail = sNode;
            }
        }

        internal void clear () {
            SNode traversal = Head;
            while (traversal != null) {
                SNode temp = traversal.Next;
                traversal.Next = null;
                traversal = temp;
            }

            Head = Tail = null;
        }

        internal void insertAt (int index, SNode sNode) {
            SNode traversalFront = Head;
            SNode traversalBack = null;

            int pointer = 0;
            while (traversalFront != null) {

                if (pointer == index) {
                    sNode.Next = traversalFront;

                    if (traversalBack != null) {
                        traversalBack.Next = sNode;
                    }

                    break;
                }

                pointer++;
                traversalBack = traversalFront;
                traversalFront = traversalFront.Next;
            }

            if (sNode.Next == Head) {
                Head = sNode;
            }
        }

        internal void removeLast () {
            if (Head == Tail) {
                Head = Tail = null;
            } else {
                SNode traversal = Head;
                while (traversal != null) {
                    if (traversal.Next == Tail) {
                        traversal.Next = null;
                        Tail = traversal;
                        break;
                    }

                    traversal = traversal.Next;
                }
            }
        }

        internal void removeFirst () {
            if (Head == Tail) {
                Head = Tail = null;
            } else {
                Head = Head.Next;
            }
        }

        internal T[] getAsArray () {
            return getDataAsEnumerable ().ToArray ();
        }

        private IEnumerable<T> getDataAsEnumerable () {
            SNode traversal = Head;
            while (traversal != null) {
                yield return traversal.Data;
                traversal = traversal.Next;
            }
        }

        public class SNode {
            public T Data { get; set; }

            public SNode Next { get; set; }

            public SNode (T data) {
                Data = data;
            }
        }
    }
}